File: src/display/model/mesh/ChunkMesh.cpp
//
// Created by Sinis on 31/05/2025.
//

#include "ChunkMesh.h"
#include "vertex/Vertex.h"
#include "../../../logs/Logs.h"

ChunkMesh::ChunkMesh(const World &world, glm::ivec3 chunkPos, uint16_t *blocks) {
    setupMesh();
    buildMesh(world, chunkPos, blocks);
}

void ChunkMesh::setupMesh() {
    glCreateBuffers(1, &VBO);
    glCreateBuffers(1, &EBO);

    glCreateVertexArrays(1, &VAO);
    glVertexArrayVertexBuffer(VAO, 0, VBO, 0, 8);
    glVertexArrayElementBuffer(VAO, EBO);

    glEnableVertexArrayAttrib(VAO, 0);
    glEnableVertexArrayAttrib(VAO, 1);

    glVertexArrayAttribFormat(VAO,0,1,GL_UNSIGNED_INT, GL_FALSE, 0);
    glVertexArrayAttribFormat(VAO,1,1,GL_UNSIGNED_INT, GL_FALSE, 4);

    glVertexArrayAttribBinding(VAO,0,0);
    glVertexArrayAttribBinding(VAO,1,0);


    printf("VBO: %u, EBO: %u, VAO: %u\n", VBO, EBO, VAO);
    Logs::log("INFO", "ChunkMesh created with VBO: " + std::to_string(VBO) + ", EBO: " + std::to_string(EBO) + ", VAO: " + std::to_string(VAO));
}

void ChunkMesh::bindData(std::vector<uint32_t> &vertices, std::vector<unsigned int> &indices) const{
    glNamedBufferData(VBO,vertices.size() * sizeof(uint32_t), &vertices[0], GL_STATIC_DRAW);
    glNamedBufferData(EBO, indices.size() * sizeof(unsigned int),&indices[0],GL_STATIC_DRAW);
}

void ChunkMesh::buildMesh(const World &world, glm::ivec3 chunkPos, const uint16_t *blocks) {
    std::vector<uint32_t> vertices;
    vertices.reserve(CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE * 6 * 4 * 2); // 6 faces, 4 vertices per face
    std::vector<unsigned int> indices;
    indices.reserve(CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE * 6 * 6); // 6 faces, 6 nbIndices per face
    int index = 0;
    for (int x = 0; x < CHUNK_SIZE; x++) {
        for (int y = 0; y < CHUNK_SIZE; y++) {
            for (int z = 0; z < CHUNK_SIZE; z++) {
                uint16_t voxel_id = blocks[x * CHUNK_SIZE * CHUNK_SIZE + y * CHUNK_SIZE + z];

                if (voxel_id == 0) continue; // skip empty blocks
                uint64_t v[4];
                //front face
                if (isVoid(glm::vec3(x, y, z + 1), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::ivec3(x, y, z + 1), glm::vec3(0.0f, 0.0f, 1.0f), 0);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z + 1), glm::vec3(0.0f, 0.0f, 1.0f), 1);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z + 1), glm::vec3(0.0f, 0.0f, 1.0f), 3);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z + 1), glm::vec3(0.0f, 0.0f, 1.0f), 2);

                    index = addData(vertices, indices, v, index);
                }
                // back face
                if (isVoid(glm::vec3(x, y, z - 1), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::ivec3(x, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 2);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 0);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 1);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 3);

                    index = addData(vertices, indices, v, index);
                }
                //top face
                if (isVoid(glm::vec3(x, y + 1, z), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 2);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 0);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 1);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 3);

                    index = addData(vertices, indices, v, index);
                }
                // bottom face
                if (isVoid(glm::vec3(x, y - 1, z), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::vec3(x, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 1);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 0);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 2);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x, y, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 3);

                    index = addData(vertices, indices, v, index);
                }

                // right face
                if (isVoid(glm::vec3(x + 1, y, z), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 2);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 3);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x + 1, y + 1, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 1);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x + 1, y, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 0);

                    index = addData(vertices, indices, v, index);
                }

                // left face
                if (isVoid(glm::vec3(x - 1, y, z), blocks, world, chunkPos)) {

                    v[0] = Vertex::packData(voxel_id, glm::vec3(x, y, z), glm::vec3(0.0f, 0.0f, -1.0f), 0);
                    v[1] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z), glm::vec3(0.0f, 0.0f, -1.0f), 1);
                    v[2] = Vertex::packData(voxel_id, glm::vec3(x, y + 1, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 3);
                    v[3] = Vertex::packData(voxel_id, glm::vec3(x, y, z + 1), glm::vec3(0.0f, 0.0f, -1.0f), 2);

                    index = addData(vertices, indices, v, index);
                }
            }
        }
    }
    this->nbIndices = indices.size();
    printf("Size %llu : %llu\n", vertices.size(), indices.size());
    bindData(vertices, indices);
}

bool ChunkMesh::isVoid(glm::ivec3 blockPos, const uint16_t *blocks, const World &world, glm::ivec3 chunkPos) {
    if (blockPos.x < 0 || blockPos.x >= CHUNK_SIZE ||
        blockPos.y < 0 || blockPos.y >= CHUNK_SIZE ||
        blockPos.z < 0 || blockPos.z >= CHUNK_SIZE) {
        return world.getBlockAt(chunkPos * CHUNK_SIZE + blockPos) == 0;
    }
    return blocks[blockPos.x * CHUNK_SIZE * CHUNK_SIZE + blockPos.y * CHUNK_SIZE + blockPos.z] == 0;
}

void ChunkMesh::draw() const {
    glBindVertexArray(VAO);
    glDrawElementsBaseVertex(GL_TRIANGLES, (int) nbIndices, GL_UNSIGNED_INT, (void *) 0, 0);
    glBindVertexArray(0);
}

int ChunkMesh::addData(std::vector<uint32_t> &vertex, std::vector<unsigned int> &indices, uint64_t *v, int index) {

    for (int i = 0; i < 4; ++i) {
        vertex.push_back((uint32_t)(v[i] >> 32));        // High 32 bits
        vertex.push_back((uint32_t)(v[i] & 0xFFFFFFFF)); // Low 32 bits
    }

    indices.push_back(index);
    indices.push_back(index + 1);
    indices.push_back(index + 2);
    indices.push_back(index);
    indices.push_back(index + 2);
    indices.push_back(index + 3);

    return index + 4;
}

File: src/client/generation/mesh/chunk_mesh.rs
use shared::print_base;
use std::collections::HashMap;
use std::sync::Arc;
use gl::types::{GLint, GLuint};
use gl::UNSIGNED_INT;
use shared::common::display::vertex::vertex::Vertex;
use shared::common::world::chunk::chunk::Chunk;
use shared::common::world::pos::chunkpos::{ChunkPos, CHUNK_SIZE};
use shared::common::world::pos::iblockpos::IBlockPos;
use shared::print_debug;
use crate::client::display::renderer::gui::text::text::MeshText;
use crate::client::display::renderer::mesh::chunk_mesh::Mesh;

pub struct ChunkMesh {
    vertices: Vec<u32>,
    indices: Vec<u32>,
    chunk_pos: ChunkPos
}


impl ChunkMesh {

    pub fn get_chunk_pos(&self) -> ChunkPos {
        self.chunk_pos
    }

    pub fn is_empty(&self) -> bool {
        self.vertices.is_empty() && self.indices.is_empty()
    }

    pub unsafe fn link(&mut self) -> Mesh {
        let (mut vao, mut vbo, mut ebo) = (0,0,0);
        Self::setup_mesh(&mut vao,&mut vbo,&mut ebo);
        self.bind_data(&mut vbo,&mut ebo);
        Mesh::new(vao,vbo,ebo, self.indices.len() as i32)
    }



    unsafe fn setup_mesh(vao: &mut GLuint, vbo: &mut GLuint, ebo: &mut GLuint) {

        // 1. Create the objects (DSA style)
        gl::CreateBuffers(1, vbo);
        gl::CreateBuffers(1, ebo);
        gl::CreateVertexArrays(1, vao);

        // 2. Configure Attribute 0
        gl::VertexArrayVertexBuffer(*vao, 0, *vbo, 0, 8);

        gl::EnableVertexArrayAttrib(*vao, 0);
        gl::EnableVertexArrayAttrib(*vao, 1);

        // 3. Configure Attribute 1
        gl::VertexArrayAttribIFormat(*vao, 0, 1, UNSIGNED_INT, 0);
        gl::VertexArrayAttribIFormat(*vao, 1, 1, UNSIGNED_INT, 4);

        gl::VertexArrayAttribBinding(*vao,0,0);
        gl::VertexArrayAttribBinding(*vao,1,0);

        gl::VertexArrayElementBuffer(*vao, *ebo);

    }

    unsafe fn bind_data(&self, vbo: &mut GLuint, ebo: &mut GLuint) {
        gl::NamedBufferData(*vbo, self.vertices.len().cast_signed() * 4, self.vertices.as_ptr() as *const _ , gl::STATIC_DRAW);
        gl::NamedBufferData(*ebo, self.indices.len().cast_signed() * 4, self.indices.as_ptr() as *const _, gl::STATIC_DRAW);
    }

    pub fn build_mesh(chunks_map : &HashMap<ChunkPos,Vec<u16>>, chunk_pos: ChunkPos) -> Option<(ChunkMesh, MeshText)> {
        let mut chunk_mesh = Self {
            vertices: Vec::new(),
            indices: Vec::new(),
            chunk_pos
        };
        let blocks = chunks_map.get(&chunk_pos)?;
        chunk_mesh.vertices.reserve(CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE * 6 * 4 * 2); // 6 faces, 4 vertices per face
        chunk_mesh.indices.reserve(CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE * 6 * 6); // 6 faces, 6 nbIndices per face
        let mut index = 0;
        for x in 0..CHUNK_SIZE {
            for y in 0..CHUNK_SIZE {
                for z in 0..CHUNK_SIZE{
                    let voxel_id : u16 = blocks[x * CHUNK_SIZE * CHUNK_SIZE + y * CHUNK_SIZE + z];

                    if voxel_id == 0 {
                        continue; // skip empty blocks
                    }
                    let mut v: [u64;4] = [0; 4];
                    let (ux, uy, uz) = (x as i32, y as i32, z as i32);
                    //front face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux, uy, uz + 1), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz + 1), 1, 3).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz + 1), 1, 2).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz + 1), 1, 0).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz + 1), 1, 1).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }
                    // back face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux, uy, uz - 1), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz), 4, 2).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz), 4, 0).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz), 4, 1).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz), 4, 3).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }
                    //top face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux, uy + 1, uz), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz), 0, 2).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz), 0, 0).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz + 1), 0, 1).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz + 1), 0, 3).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }
                    // bottom face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux, uy - 1, uz), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz), 5, 1).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz + 1), 5, 3).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz + 1), 5, 2).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz), 5, 0).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }

                    // right face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux + 1, uy, uz), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz), 2, 1).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy, uz + 1), 2, 3).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz + 1), 2, 2).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux + 1, uy + 1, uz), 2, 0).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }

                    // left face
                    if chunk_mesh.is_void(IBlockPos::from_ints(ux - 1, uy, uz), blocks, chunks_map, chunk_pos) {

                        v[0] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz), 3, 3).unwrap();
                        v[1] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz), 3, 2).unwrap();
                        v[2] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy + 1, uz + 1), 3, 0).unwrap();
                        v[3] = Vertex::pack_data(voxel_id, glam::IVec3::new(ux, uy, uz + 1), 3, 1).unwrap();

                        index = chunk_mesh.add_data(v, index);
                    }
                }
            }
        }
        if chunk_mesh.indices.len() == 0 {
            return None
        };
        print_debug!("Created {} vertices in chunks", chunk_mesh.indices.len());
        let mesh_text = unsafe { MeshText::new(chunk_mesh.vertices.clone(), chunk_mesh.indices.clone()) };
        Some((chunk_mesh, mesh_text))
    }

    fn is_void(&self, block_pos: IBlockPos, blocks : &Vec<u16>, chunks_map : &HashMap<ChunkPos,Vec<u16>>, chunk_pos: ChunkPos) -> bool {
        if block_pos.x < 0 || block_pos.x >= CHUNK_SIZE as i32 ||
            block_pos.y < 0 || block_pos.y >= CHUNK_SIZE as i32 ||
            block_pos.z < 0 || block_pos.z >= CHUNK_SIZE as i32 {

            return self.get_block_at(chunk_pos + block_pos, chunks_map) == 0;
        }
        blocks[block_pos.x as usize * CHUNK_SIZE * CHUNK_SIZE + block_pos.y as usize * CHUNK_SIZE + block_pos.z as usize] == 0
    }

    pub fn get_block_at(&self, ipos : IBlockPos, chunks_map : &HashMap<ChunkPos,Vec<u16>>,) -> u16 {
        let sz = CHUNK_SIZE as i32;

        let (block_pos, chunk_pos) = ipos.as_split_pos();
        if chunk_pos.y < -2 || chunk_pos.y > 9 {
            return 0
        }
        chunks_map.get(&chunk_pos).unwrap()[block_pos.get_offset()]
    }

    fn add_data(&mut self, v : [u64;4], index : u32) -> u32 {

        for i in 0..4usize {
            self.vertices.push((v[i] >> 32) as u32);        // High 32 bits
            self.vertices.push((v[i] & 0xFFFFFFFF) as u32); // Low 32 bits
        }

        self.indices.push(index);
        self.indices.push(index + 2);
        self.indices.push(index + 1);
        self.indices.push(index);
        self.indices.push(index + 3);
        self.indices.push(index + 2);

        index + 4
    }
}